core/src/test/java/io/grpc/internal/SpiffeUtilTest.java

```diff
@@ -193,4 +210,184 @@ public void spiffePathFormatTest() {
           + "underscores ([a-zA-Z0-9.-_])", iae.getMessage());
     }
   }
+
+  public static class CertificateApiTest {
+    private static final String SPIFFE_PEM_FILE = "spiffe_cert.pem";
+    private static final String MULTI_URI_SAN_PEM_FILE = "spiffe_multi_uri_san_cert.pem";
+    private static final String SERVER_0_PEM_FILE = "server0.pem";
+    private static final String TEST_DIRECTORY_PREFIX = "io/grpc/internal/";
+    private static final String SPIFFE_TRUST_BUNDLE = "spiffebundle.json";
+    private static final String SPIFFE_TRUST_BUNDLE_MALFORMED = "spiffebundle_malformed.json";
+    private static final String SPIFFE_TRUST_BUNDLE_CORRUPTED_CERT =
+        "spiffebundle_corrupted_cert.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_KTY = "spiffebundle_wrong_kty.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_KID = "spiffebundle_wrong_kid.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_USE = "spiffebundle_wrong_use.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_MULTI_CERTS =
+        "spiffebundle_wrong_multi_certs.json";
+    private static final String SPIFFE_TRUST_BUNDLE_DUPLICATES = "spiffebundle_duplicates.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_ROOT = "spiffebundle_wrong_root.json";
+    private static final String SPIFFE_TRUST_BUNDLE_WRONG_SEQ = "spiffebundle_wrong_seq_type.json";
+    private static final String DOMAIN_ERROR_MESSAGE =
+        " Certificate loading for trust domain 'google.com' failed.";
+
+
+    private final TemporaryFolder tempFolder = new TemporaryFolder();
+    private String tempSpiffeTrustBundle;
+    private String tempSpiffeTrustBundleMalformed;
+    private String tempSpiffeTrustBundleCorruptedCert;
+    private String tempSpiffeTrustBundleWrongKty;
+    private String tempSpiffeTrustBundleWrongKid;
+    private String tempSpiffeTrustBundleWrongUse;
+    private String tempSpiffeTrustBundleWrongMultiCerts;
+    private String tempSpiffeTrustBundleDuplicates;
+    private String tempSpiffeTrustBundleWrongRoot;
+    private String tempSpiffeTrustBundleWrongSeq;
+
+
+    private X509Certificate[] spiffeCert;
+    private X509Certificate[] multipleUriSanCert;
+    private X509Certificate[] serverCert0;
+
+    @Before
+    public void setUp() throws Exception {
+      tempFolder.create();
+      spiffeCert = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SPIFFE_PEM_FILE));
+      multipleUriSanCert = CertificateUtils.getX509Certificates(TlsTesting
+          .loadCert(MULTI_URI_SAN_PEM_FILE));
+      serverCert0 = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SERVER_0_PEM_FILE));
+
+      tempSpiffeTrustBundle = copyFileToTmp(SPIFFE_TRUST_BUNDLE);
```

Comment:
You didn't want to inline these to where they are used? Since you are deleting the entire folder, you don't need to track each in a variable. And each of these is only used once, so at the very least make it a local variable.

```java
SpiffeBundle tb = SpiffeUtil.loadTrustBundleFromFile(tempSpiffeTrustBundle);
// vs
String tempSpiffeTrustBundle = copyFileToTmp(SPIFFE_TRUST_BUNDLE);
SpiffeBundle tb = SpiffeUtil.loadTrustBundleFromFile(tempSpiffeTrustBundle);
// or
SpiffeBundle tb = SpiffeUtil.loadTrustBundleFromFile(copyFileToTmp(SPIFFE_TRUST_BUNDLE));
```